cpp/ql/lib/semmle/code/cpp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -0,0 +1,248 @@
+/**
+ * Provides classes and predicates for defining flow summaries.
+ */
+
+private import cpp
+private import codeql.dataflow.internal.FlowSummaryImpl
+private import codeql.dataflow.internal.AccessPathSyntax as AccessPath
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowPrivate
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowUtil
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowImplSpecific as DataFlowImplSpecific
+private import semmle.code.cpp.dataflow.ExternalFlow
+
+module Input implements InputSig<DataFlowImplSpecific::CppDataFlow> {
+  class SummarizedCallableBase = Function;
+
+  ArgumentPosition callbackSelfParameterPosition() { result = TDirectPosition(-1) }
+
+  ReturnKind getStandardReturnValueKind() { result.(NormalReturnKind).getIndirectionIndex() = 0 }
+
+  string encodeParameterPosition(ParameterPosition pos) { result = pos.toString() }
+
+  string encodeArgumentPosition(ArgumentPosition pos) { result = pos.toString() }
+
+  string encodeReturn(ReturnKind rk, string arg) {
+    rk != getStandardReturnValueKind() and
+    result = "ReturnValue" and
+    arg = rk.toString()
+  }
+
+  string encodeContent(ContentSet cs, string arg) {
+    exists(FieldContent c |
+      cs.isSingleton(c) and
+      result = "Field" and
+      arg = c.getField().getName()
+    )
```

Comment:
A `FieldContent` is defined by a field name and an indirection index. So eventually this will need to include the indirection index in the syntax somehow - either through the `arg` column or the `result` column. Right now I think we're just conflating all indirections of a field when using `encodeContent`) which is probably fine for now 👍

Reply:
Yes we are conflating indirections, and it's definitely something I want to fix (though not necessarily in this PR - as you say it's probably fine for now).  I hadn't realized indirection is built into `FieldContent` though - why is that?  I was imagining it as perhaps its own content type, so you might have an access path something like `Argument[0].fieldName.Indirect[1]` or `ReturnValue.Indirect[2]`.

Reply:
A `FieldContent` is defined as a pair consisting of a field and an indirection index because a read of a FieldContent `(field, n)` corresponds to flow through an `FieldAddressInstruction` that's associated with `field`, followed by flow through `n` `LoadInstruction`s (see [here](https://github.com/github/codeql/blob/f1d6f56621409ea705438bf189f15fd617bc16e1/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll#L822)).

Other languages have both `FieldContent` and `ArrayContent` which splits out this into two contents, but when we tried to do in C/C++ a long time ago and gave huge performance problems because of the way the shared dataflow library calculated the list of successor contents that could follow an `ArrayContent`.

Reply:
> I was imagining it as perhaps its own content type, so you might have an access path something like Argument[0].fieldName.Indirect[1] or ReturnValue.Indirect[2].

We could probably implement that syntax, but the syntax that would naturally follow from the way we've designed the C/C++ content-based flow is `Argument[0].fieldName[1]` (or something like `Argument[0].*fieldName` which I'd probably prefer since it looks more like C/C++ syntax)

Reply:
We'll need the same syntax for return values (e.g. `*ReturnValue`) and for non-field arguments / parameters, like maybe `*Argument[0]`.  Then there's the difference between `Argument[0].*fieldName` and `*Argument[0].fieldName`.  It might get a bit fiddly, but if indirections are robustly supported at the dataflow level this should all be do-able.

Reply:
> We'll need the same syntax for return values (e.g. `*ReturnValue`) and for non-field arguments / parameters, like maybe `*Argument[0]`. Then there's the difference between `Argument[0].*fieldName` and `*Argument[0].fieldName`.

Indeed, `Argument[0].*fieldName` would be "the value pointed to by the `fieldName` field of the 0'th argument" and `*Argument[0].fieldName` would be "the value of `fieldName` that you can get to by dereferencing the 0'th argument".

> It might get a bit fiddly, but if indirections are robustly supported at the dataflow level this should all be do-able.

I really hope so 😂. That is certainly my current belief 🤞.

Reply:
@MathiasVP I'm ready to discuss indirection syntax in more detail.  Lets say we want to reference the `first` field of the argument `pair` in a call to:
```
void myFunc(MyPair *pair)
```
The syntax for this could be:
```
"explicit indirection"
Argument[0].Indirect[1].first

"argument list"
Argument[0,1].first

"C-style"
*Argument[0].first
```
... or anything else we can come up with.  I'm currently leaning towards the third one ("c-style"), which you seemed to like the look of above.  I'm a bit worried that the non-standardness of this syntax compared to other languages will cause problems though, e.g. arguments and parameters appear to be decoded more-or-less in shared code (thus with less flexibility).  Decoding `Argument[*0].first` might be a lot easier to implement, but inconsistent with `*ReturnValue` for example.

Whichever way we do it I think there will have to be explicitly coded cases supporting each of: indirect arguments, return values, variable accesses, field accesses.  Also union content if/when we add that.  Possibly a special case for post-update nodes if required.  I think most of the work will be done in the `encodeContent`, `encodeReturn` and similar predicates.

Does the above all sound reasonable to you?

Reply:
That all sounds right, yes. I'm also very much in favor of the third option ("c-style") since this is also the output currently used on `toString`s on dataflow nodes (so that's what users are seeing in path explanations already), and I think there's value in keeping those things in sync.

It's also worth considering what to do if `first` is a pointer, and we want to model that the function is reading what `first` points to. I guess the syntax should then be `*Argument[0].*first`

Reply:
Yeah, I encountered that possibility while writing test cases.  I think it will work out if we support `*` everywhere we should.  Trickier still is a less common case, a source that is the indirection of a global variable:
```
int *globalIndirectSource;
```
For which the syntax to make that a source is currently supposed to be something like:
```
";;false;globalIndirectSource;;;*;remote"
```
Which is ugly, and I'm not sure if I can make it work.

Reply:
(indirect globals is now a follow-up issue)